File: Behavior/Visitor/Go/visitor.go
package main

import "fmt"

// Element 接口
type Element interface {
	Accept(visitor Visitor)
}

// Character 结构体
type Character struct{}

func (c *Character) Accept(visitor Visitor) {
	visitor.VisitCharacter(c)
}

// Enemy 结构体
type Enemy struct{}

func (e *Enemy) Accept(visitor Visitor) {
	visitor.VisitEnemy(e)
}

// Item 结构体
type Item struct{}

func (i *Item) Accept(visitor Visitor) {
	visitor.VisitItem(i)
}

// Visitor 接口
type Visitor interface {
	VisitCharacter(*Character)
	VisitEnemy(*Enemy)
	VisitItem(*Item)
}

// RenderVisitor 结构体实现了 Visitor 接口
type RenderVisitor struct{}

func (rv *RenderVisitor) VisitCharacter(c *Character) {
	fmt.Println("Rendering a Character")
}

func (rv *RenderVisitor) VisitEnemy(e *Enemy) {
	fmt.Println("Rendering an Enemy")
}

func (rv *RenderVisitor) VisitItem(i *Item) {
	fmt.Println("Rendering an Item")
}

func main() {
	elements := []Element{&Character{}, &Enemy{}, &Item{}}

	renderVisitor := &RenderVisitor{}

	for _, element := range elements {
		element.Accept(renderVisitor)
	}
}

File: Behavior/Visitor/TypeScript/visitor.ts
// 元素接口
interface Element {
    accept(visitor: Visitor): void;
}

// 角色类
class Character implements Element {
    accept(visitor: Visitor): void {
        visitor.visitCharacter(this);
    }

    // 角色特定的方法
    getCharacterDetails(): string {
        return 'Character details...';
    }
}

// 敌人类
class Enemy implements Element {
    accept(visitor: Visitor): void {
        visitor.visitEnemy(this);
    }

    // 敌人特定的方法
    getEnemyDetails(): string {
        return 'Enemy details...';
    }
}

// 道具类
class Item implements Element {
    accept(visitor: Visitor): void {
        visitor.visitItem(this);
    }

    // 道具特定的方法
    getItemDetails(): string {
        return 'Item details...';
    }
}

// 访问者接口
interface Visitor {
    visitCharacter(character: Character): void;
    visitEnemy(enemy: Enemy): void;
    visitItem(item: Item): void;
}

// 调试和测试访问者
class DebugVisitor implements Visitor {
    visitCharacter(character: Character): void {
        console.log('Debugging Character: ' + character.getCharacterDetails());
    }

    visitEnemy(enemy: Enemy): void {
        console.log('Debugging Enemy: ' + enemy.getEnemyDetails());
    }

    visitItem(item: Item): void {
        console.log('Debugging Item: ' + item.getItemDetails());
    }
}

function main(): void {
    const elements: Element[] = [new Character(), new Enemy(), new Item()];

    const debugVisitor = new DebugVisitor();

    elements.forEach(element => {
        element.accept(debugVisitor);
    });
}

main();
